src/algokit_utils/_debug_utils.py

```diff
@@ -0,0 +1,212 @@
+    program_hash: str = field(metadata={"json": "hash"})
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, AVMDebuggerSourceMapEntry):
+            return self.location == other.location and self.program_hash == other.program_hash
+        return False
+
+    def __str__(self) -> str:
+        return json.dumps({"sourcemap-location": self.location, "hash": self.program_hash})
+
+
+@dataclass
+class AVMDebuggerSourceMap:
+    txn_group_sources: list[AVMDebuggerSourceMapEntry] = field(metadata={"json": "txn-group-sources"})
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "AVMDebuggerSourceMap":
+        return cls(
+            txn_group_sources=[
+                AVMDebuggerSourceMapEntry(location=item["sourcemap-location"], program_hash=item["hash"])
+                for item in data.get("txn-group-sources", [])
+            ]
+        )
+
+    def to_dict(self) -> dict:
+        return {"txn-group-sources": [json.loads(str(item)) for item in self.txn_group_sources]}
+
+
+@dataclass
+class PersistSourceMapInput:
+    file_name: str
+    teal: str
+    app_name: str
+
+
+def _load_or_create_sources(project_root: Path) -> AVMDebuggerSourceMap:
+    sources_path = project_root / ALGOKIT_DIR / SOURCES_DIR / SOURCES_FILE
+    if not sources_path.exists():
+        return AVMDebuggerSourceMap(txn_group_sources=[])
+
+    with sources_path.open() as f:
+        return AVMDebuggerSourceMap.from_dict(json.load(f))
+
+
+def _upsert_debug_sourcemaps(sourcemaps: list[AVMDebuggerSourceMapEntry], project_root: Path) -> None:
+    sources_path = project_root / ALGOKIT_DIR / SOURCES_DIR / SOURCES_FILE
+    sources = _load_or_create_sources(sources_path)
+
+    for sourcemap in sourcemaps:
+        if sourcemap not in sources.txn_group_sources:
+            sources.txn_group_sources.append(sourcemap)
+        else:
+            index = sources.txn_group_sources.index(sourcemap)
+            sources.txn_group_sources[index] = sourcemap
+
+    with sources_path.open("w") as f:
+        json.dump(sources.to_dict(), f)
+
+
+def _write_to_file(path: Path, content: str) -> None:
+    path.parent.mkdir(parents=True, exist_ok=True)
+    path.write_text(content)
+
+
+def _build_avm_sourcemap(
+    teal_content: str, app_name: str, file_name: str, output_path: Path, client: "AlgodClient"
+) -> AVMDebuggerSourceMapEntry:
+    file_name = f"{file_name}{TEAL_FILE_EXT}" if not file_name.endswith(TEAL_FILE_EXT) else file_name
+    result = client.compile(deploy.strip_comments(teal_content), source_map=True)
+    program_hash = base64.b64encode(
+        checksum(base64.b64decode(result["result"]))  # type: ignore[no-untyped-call]
+    ).decode()
+    source_map = SourceMap(result["sourcemap"]).__dict__
+    source_map["sources"] = [file_name]
+
+    output_dir_path = output_path / ALGOKIT_DIR / SOURCES_DIR / app_name
+    source_map_output_path = output_dir_path / f'{file_name.replace(TEAL_FILE_EXT, "")}{TEAL_SOURCEMAP_EXT}'
```

Comment:
You could keep `TEAL_SOURCEMAP_EXT = ".tok.map"`, change f'{file_name.replace(TEAL_FILE_EXT, "")}{TEAL_SOURCEMAP_EXT}' --> f'{file_name}{TEAL_SOURCEMAP_EXT}'

Reply:
@neilcampbell instead using a custom setter on the abstraction itself so it strips down word .teal if it ends with that during set, remaining functions can safely assume its a string that doesnt end with word .teal